src/read/xcoff/symbol.rs

```diff
@@ -0,0 +1,429 @@
+pub struct XcoffSymbolIterator<'data, 'file, Xcoff, R = &'data [u8]>
+where
+    'data: 'file,
+    Xcoff: FileHeader,
+    R: ReadRef<'data>,
+{
+    pub(crate) file: &'file XcoffFile<'data, Xcoff, R>,
+    pub(super) symbols: &'file SymbolTable<'data, Xcoff, R>,
+    pub(super) index: usize,
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> fmt::Debug
+    for XcoffSymbolIterator<'data, 'file, Xcoff, R>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XcoffSymbolIterator").finish()
+    }
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> Iterator
+    for XcoffSymbolIterator<'data, 'file, Xcoff, R>
+{
+    type Item = XcoffSymbol<'data, 'file, Xcoff, R>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.index;
+        let symbol = self.symbols.symbols.get(index)?;
+        self.index += 1;
+        Some(XcoffSymbol {
+            file: self.file,
+            symbols: self.symbols,
+            index: SymbolIndex(index),
+            symbol,
+        })
+    }
+}
+
+/// A symbol of an `XcoffFile32`.
+pub type XcoffSymbol32<'data, 'file, R = &'data [u8]> =
+    XcoffSymbol<'data, 'file, xcoff::FileHeader32, R>;
+/// A symbol of an `XcoffFile64`.
+pub type XcoffSymbol64<'data, 'file, R = &'data [u8]> =
+    XcoffSymbol<'data, 'file, xcoff::FileHeader64, R>;
+
+/// A symbol of an `XcoffFile`.
+#[derive(Debug, Clone, Copy)]
+pub struct XcoffSymbol<'data, 'file, Xcoff, R = &'data [u8]>
+where
+    'data: 'file,
+    Xcoff: FileHeader,
+    R: ReadRef<'data>,
+{
+    pub(crate) file: &'file XcoffFile<'data, Xcoff, R>,
+    pub(super) symbols: &'file SymbolTable<'data, Xcoff, R>,
+    pub(super) index: SymbolIndex,
+    pub(super) symbol: &'data Xcoff::Symbol,
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> read::private::Sealed
+    for XcoffSymbol<'data, 'file, Xcoff, R>
+{
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> ObjectSymbol<'data>
+    for XcoffSymbol<'data, 'file, Xcoff, R>
+{
+    #[inline]
+    fn index(&self) -> SymbolIndex {
+        self.index
+    }
+
+    fn name_bytes(&self) -> Result<&'data [u8]> {
+        self.symbol.name(self.symbols.strings)
+    }
+
+    fn name(&self) -> Result<&'data str> {
+        let name = self.name_bytes()?;
+        str::from_utf8(name)
+            .ok()
+            .read_error("Non UTF-8 XCOFF symbol name")
+    }
+
+    #[inline]
+    fn address(&self) -> u64 {
+        return self.symbol.n_value().into();
+    }
+
+    #[inline]
+    fn size(&self) -> u64 {
+        // TODO: get the symbol size when the csect auxiliary symbol is supported.
+        // Most symbols don't have sizes.
+        0
+    }
+
+    fn kind(&self) -> SymbolKind {
+        match self.symbol.n_sclass() {
+            xcoff::C_FILE => SymbolKind::File,
+            xcoff::C_NULL => SymbolKind::Null,
+            xcoff::C_GTLS | xcoff::C_STTLS => SymbolKind::Tls,
```

Comment:
I'm not sure about this. They are debug symbols that follow the actual symbol, and those actual symbols should be handled below already by looking at the section kind.

Reply:
My idea is that if we can directly deduce the symbol type from the storage class of the current symbol, then we will use it, otherwise we will refer to the kind of the section it corresponds to.

Reply:
But these aren't the actual symbols, so I suspect they shouldn't have this kind. Is there an example of these symbols in the test files you added?